File: src/services/email.ts
import nodemailer from 'nodemailer';
import Handlebars from 'handlebars';
import { log } from '../utils/logger';

const SMTP_HOST = process.env.SMTP_HOST;
const SMTP_PORT = parseInt(process.env.SMTP_PORT || '587');
const SMTP_USER = process.env.SMTP_USER;
const SMTP_PASS = process.env.SMTP_PASS;
const SMTP_FROM = process.env.SMTP_FROM || 'noreply@foundryvtt-relay.com';
const SMTP_SECURE = process.env.SMTP_SECURE === 'true';
const FRONTEND_URL = process.env.FRONTEND_URL || 'http://localhost:3010';

let transporter: nodemailer.Transporter | null = null;

if (SMTP_HOST) {
  transporter = nodemailer.createTransport({
    host: SMTP_HOST,
    port: SMTP_PORT,
    secure: SMTP_SECURE,
    auth: (SMTP_USER && SMTP_PASS) ? {
      user: SMTP_USER,
      pass: SMTP_PASS
    } : undefined
  });
}

const resetEmailTemplate = Handlebars.compile(`
<!DOCTYPE html>
<html>
<head>
  <style>
    body { font-family: Arial, sans-serif; background: #f4f4f4; padding: 20px; }
    .container { max-width: 600px; margin: 0 auto; background: white; border-radius: 8px; padding: 30px; }
    .btn { display: inline-block; background: #4a6fa5; color: white; padding: 12px 24px; text-decoration: none; border-radius: 4px; margin: 20px 0; }
    .footer { margin-top: 30px; font-size: 12px; color: #999; }
  </style>
</head>
<body>
  <div class="container">
    <h2>Password Reset Request</h2>
    <p>You requested a password reset for your Foundry REST API account.</p>
    <p>Click the button below to reset your password. This link expires in 1 hour.</p>
    <a href="{{resetUrl}}" class="btn">Reset Password</a>
    <p>If the button doesn't work, copy and paste this URL into your browser:</p>
    <p style="word-break: break-all; font-size: 14px; color: #666;">{{resetUrl}}</p>
    <div class="footer">
      <p>If you didn't request this, you can safely ignore this email.</p>
      <p>Foundry REST API Relay</p>
    </div>
  </div>
</body>
</html>
`);

export async function sendPasswordResetEmail(to: string, resetToken: string): Promise<void> {
  const resetUrl = `${FRONTEND_URL}?reset-token=${resetToken}`;

  if (!transporter) {
    log.info('SMTP not configured — password reset URL (for development):');
    log.info(resetUrl);
    return;
  }

  try {
    const html = resetEmailTemplate({ resetUrl });

    await transporter.sendMail({
      from: SMTP_FROM,
      to,
      subject: 'Password Reset — Foundry REST API',
      html
    });

    log.info('Password reset email sent', { to });
  } catch (error) {
    log.error('Failed to send password reset email', { error, to });
  }
}

File: go-relay/internal/service/email.go
package service

import (
	"fmt"
	"net/mail"
	"net/smtp"
	"strings"
	"time"

	"github.com/ThreeHats/foundryvtt-rest-api-relay/go-relay/internal/config"
	"github.com/rs/zerolog/log"
)

// extractEmailAddress parses a "Name <addr>" string and returns just the addr.
// If parsing fails, returns the input as-is.
func extractEmailAddress(from string) string {
	addr, err := mail.ParseAddress(from)
	if err != nil {
		return from
	}
	return addr.Address
}

// SendVerificationEmail sends an email verification email, or logs the URL if SMTP is not configured.
func SendVerificationEmail(cfg *config.Config, to, verifyToken string) {
	verifyURL := fmt.Sprintf("%s/auth/verify?token=%s", cfg.FrontendURL, verifyToken)

	if cfg.SMTPHost == "" {
		log.Info().Str("email", to).Str("verifyURL", verifyURL).Msg("Email verification (SMTP not configured)")
		return
	}

	subject := "Verify Your Email Address"
	body := fmt.Sprintf(`<!DOCTYPE html>
<html>
<body style="font-family: Arial, sans-serif; max-width: 600px; margin: 0 auto;">
  <h2>Verify Your Email</h2>
  <p>Thanks for registering! Please verify your email address by clicking the link below:</p>
  <p><a href="%s" style="background: #4CAF50; color: white; padding: 10px 20px; text-decoration: none; border-radius: 4px;">Verify Email</a></p>
  <p>This link expires in 24 hours.</p>
  <p>If you didn't create an account, you can safely ignore this email.</p>
</body>
</html>`, verifyURL)

	msg := strings.Join([]string{
		fmt.Sprintf("From: %s", cfg.SMTPFrom),
		fmt.Sprintf("To: %s", to),
		fmt.Sprintf("Subject: %s", subject),
		"MIME-Version: 1.0",
		"Content-Type: text/html; charset=UTF-8",
		"",
		body,
	}, "\r\n")

	addr := fmt.Sprintf("%s:%d", cfg.SMTPHost, cfg.SMTPPort)
	var auth smtp.Auth
	if cfg.SMTPUser != "" {
		auth = smtp.PlainAuth("", cfg.SMTPUser, cfg.SMTPPass, cfg.SMTPHost)
	}

	go func() {
		var lastErr error
		for attempt := 0; attempt < 3; attempt++ {
			if attempt > 0 {
				time.Sleep(time.Duration(attempt*2) * time.Second)
			}
			if err := smtp.SendMail(addr, auth, extractEmailAddress(cfg.SMTPFrom), []string{to}, []byte(msg)); err != nil {
				lastErr = err
				log.Warn().Err(err).Int("attempt", attempt+1).Str("to", to).Msg("Failed to send verification email, retrying...")
				continue
			}
			log.Info().Str("to", to).Msg("Verification email sent")
			return
		}
		log.Error().Err(lastErr).Str("to", to).Msg("Failed to send verification email after 3 attempts")
	}()
}

// SendPasswordResetEmail sends a password reset email, or logs the URL if SMTP is not configured.
func SendPasswordResetEmail(cfg *config.Config, to, resetToken string) {
	if cfg.SMTPHost == "" {
		// Fallback: log the reset URL
		resetURL := fmt.Sprintf("%s?reset-token=%s", cfg.FrontendURL, resetToken)
		log.Info().Str("email", to).Str("resetURL", resetURL).Msg("Password reset (SMTP not configured)")
		return
	}

	resetURL := fmt.Sprintf("%s?reset-token=%s", cfg.FrontendURL, resetToken)

	subject := "Password Reset Request"
	body := fmt.Sprintf(`<!DOCTYPE html>
<html>
<body style="font-family: Arial, sans-serif; max-width: 600px; margin: 0 auto;">
  <h2>Password Reset</h2>
  <p>You requested a password reset. Click the link below to reset your password:</p>
  <p><a href="%s" style="background: #4CAF50; color: white; padding: 10px 20px; text-decoration: none; border-radius: 4px;">Reset Password</a></p>
  <p>This link expires in 1 hour.</p>
  <p>If you didn't request this, you can safely ignore this email.</p>
</body>
</html>`, resetURL)

	msg := strings.Join([]string{
		fmt.Sprintf("From: %s", cfg.SMTPFrom),
		fmt.Sprintf("To: %s", to),
		fmt.Sprintf("Subject: %s", subject),
		"MIME-Version: 1.0",
		"Content-Type: text/html; charset=UTF-8",
		"",
		body,
	}, "\r\n")

	addr := fmt.Sprintf("%s:%d", cfg.SMTPHost, cfg.SMTPPort)
	var auth smtp.Auth
	if cfg.SMTPUser != "" {
		auth = smtp.PlainAuth("", cfg.SMTPUser, cfg.SMTPPass, cfg.SMTPHost)
	}

	go func() {
		var lastErr error
		for attempt := 0; attempt < 3; attempt++ {
			if attempt > 0 {
				time.Sleep(time.Duration(attempt*2) * time.Second) // 2s, 4s backoff
			}
			if err := smtp.SendMail(addr, auth, extractEmailAddress(cfg.SMTPFrom), []string{to}, []byte(msg)); err != nil {
				lastErr = err
				log.Warn().Err(err).Int("attempt", attempt+1).Str("to", to).Msg("Failed to send password reset email, retrying...")
				continue
			}
			log.Info().Str("to", to).Msg("Password reset email sent")
			return
		}
		log.Error().Err(lastErr).Str("to", to).Msg("Failed to send password reset email after 3 attempts")
	}()
}
